### DISCUSSION TITLE
Maintaining backwards compatibility with WESTPA-1.0

### DISCUSSION THREAD

**Discussion by jdrusso** [westpa/westpa#211] [Ideas]
Recently I've been using the (AWESOME) `westpa.analysis` module, and I noticed an issue with backwards compatibility cropping up with respect to loading in bin mappers from WESTPA-1.0 simulation data. (Note that this isn't specifically an issue with `westpa.analysis`, that's just the context I encountered this in)

In particular, if you try to load a bin mapper from a WESTPA-1.0 simulation using WESTPA-2.0 `westpa.analysis`, it fails with
```
Traceback (most recent call last):
   <snip>
  File "/home/groups/ZuckermanLab/russojd/msm_we/msm_we/msm_we.py", line 2834, in cluster_stratified
    bin_mapper = iteration.bin_mapper
  File "/home/groups/ZuckermanLab/russojd/miniconda/envs/py38WEenv/lib/python3.8/functools.py", line 967, in __get__
    val = self.func(instance)
  File "/home/groups/ZuckermanLab/russojd/westpa/westpa/src/westpa/analysis/core.py", line 196, in bin_mapper
    mapper, _, _ = westpa.tools.binning.mapper_from_hdf5(self.run.h5file['bin_topologies'], self.h5group.attrs['binhash'])
  File "/home/groups/ZuckermanLab/russojd/westpa/westpa/src/westpa/tools/binning.py", line 82, in mapper_from_hdf5
    mapper = pickle.loads(pkldat)
ModuleNotFoundError: No module named 'westpa.binning'
``` 
because the pickled binmapper in `west.h5` is a `westpa.binning` object, but in WESTPA-2.0, that's now namespaced under `westpa.core.binning`.

On one hand, we could just say that analyzing WESTPA-1.0 with 2.0's analysis is just not supported. However, I was messing around with this and came up with a (admittedly, pretty kludgey) workaround, if we're interested in maintaining that. In `westpa/tools/binning.py`, you can add at line 82 something like
```
import westpa.core.binning, sys
sys.modules['westpa.binning'] = sys.modules['westpa.core.binning']
```
which will import `westpa.core.binning`, and 'alias' it so-to-speak under `westpa.binning`. This appears to allow you unpickle westpa-1.0 bin mappers..

You could also imagine a maybe smarter version that wraps the unpickle in a `try`, catches an `ImportError` and attempts the above, and then if that works, prints out a notice about loading old-style data or something. That way the `westpa.binning` alias isn't created unless absolutely necessary.

So, open question: Do we care enough about maintaining backwards compatibility to include a (hacky) workaround like this? If so, are there other places that would benefit from something similar? Would appreciate any feedback on this!

**Comment by synapticarbors** [comment]
I mentioned a similar solution in https://github.com/westpa/westpa/pull/106#issuecomment-636064631 early on when I was reorganizing the code. I don't think it's the worst work-around, but I don't love it either. I'm not sure if there is a better way to do this. I also wonder if there's a better serialization protocol this we could use that wouldn't tie data in the h5 files to the organization of the westpa module layout.

**Reply by jdrusso** [reply]
Thanks for linking that issue! Looks like I arrived at the same workaround you did, with the same feeling about it. 

Interesting thought about better serialization protocol. I could see that being straightforward for the rectilinear binmapper, where you really just need to store the boundaries, but I'm not familiar enough with the others to know if something hairy would crop up. I'd defer to someone more experienced with the others than me on that.

In my opinion, it would be a nice side-benefit to have the bin bounds stored in a human-readable format, vs pickled/encoded. But, again, that might not be portable to other bin mappers.

**Reply by jpthompson17** [reply]
I got curious about this and started playing around with YAML (re)serialization of bin mappers: https://github.com/jpthompson17/westpa/blob/8e90f848343776433fd88ca23eb850b8069b85f5/src/westpa/core/binning/assign.py#L144
Sharing here in case anyone is interested.

**Comment by mczwier** [comment]
The whole bin mapping system is so flexible that you would need the
equivalent of complete Python serialization. There are ways to modify
serialization so that it doesn't depend on module layout, or becomes (more)
human-readable. You could also special-case common cases, but I generally
avoid that like the plague, because (like the plague) it tends to spread in
unpredictable and sickening ways over time.

**Comment by jpthompson17** [comment]
Circling back to this, I think we could write a custom [unpickler](https://docs.python.org/3/library/pickle.html#pickle.Unpickler) and override its `find_class()` method to work around the namespace issue. Is this still worth doing? I'm all for restoring some backward-compatibility. Going forward, though, it might be easier to avoid storing any pickled data in the HDF5 file.
